iceoryx_utils/include/iceoryx_utils/internal/cxx/storable_function.hpp

```diff
@@ -0,0 +1,165 @@
+// limitations under the License.
+
+#ifndef IOX_UTILS_STORABLE_FUNCTION_HPP
+#define IOX_UTILS_STORABLE_FUNCTION_HPP
+
+#include "iceoryx_utils/cxx/function_ref.hpp"
+#include "iceoryx_utils/cxx/type_traits.hpp"
+#include "iceoryx_utils/internal/cxx/static_storage.hpp"
+
+#include <iostream>
+#include <type_traits>
+#include <utility>
+
+namespace iox
+{
+namespace cxx
+{
+template <typename ReturnType, typename... Args>
+using signature = ReturnType(Args...);
+
+template <typename StorageType, typename T>
+class storable_function;
+
+/// @brief A storable alternative of std::function which not necessarily uses dynamic memory.
+/// @param StorageType The type of internal storage to store the actual data.
+///                    Needs to provide allocate and deallocate functions.
+///                    See static_storage.hpp for a static memory version.
+/// @param ReturnType  The return type of the stored callable.
+/// @param Args        The arguments of the stored callable.
+template <typename StorageType, typename ReturnType, typename... Args>
+class storable_function<StorageType, signature<ReturnType, Args...>>
+{
+  public:
+    using signature_t = signature<ReturnType, Args...>;
+
+    storable_function() noexcept = default;
+
+    /// @brief construct from functor (including lambda)
+    template <typename Functor,
+              typename = typename std::enable_if<std::is_class<Functor>::value
+                                                     && is_invocable_r<ReturnType, Functor, Args...>::value,
+                                                 void>::type>
+    storable_function(const Functor& functor) noexcept;
+
+    /// @brief construct from function pointer (including static functions)
+    storable_function(ReturnType (*function)(Args...)) noexcept;
+
+    /// @brief construct from object reference and member function
+    /// only a pointer to the object is stored for the call
+    template <typename T, typename = typename std::enable_if<std::is_class<T>::value, void>::type>
+    storable_function(T& object, ReturnType (T::*method)(Args...)) noexcept;
+
+    /// @brief construct from object reference and const member function
+    /// only a pointer to the object is stored for the call
+    template <typename T, typename = typename std::enable_if<std::is_class<T>::value, void>::type>
+    storable_function(const T& object, ReturnType (T::*method)(Args...) const) noexcept;
+
+    storable_function(const storable_function& other) noexcept;
+
+    storable_function(storable_function&& other) noexcept;
+
+    storable_function& operator=(const storable_function& rhs) noexcept;
+
+    storable_function& operator=(storable_function&& rhs) noexcept;
+
+    ~storable_function() noexcept;
+
+    /// @brief invoke the stored function
+    /// @note  invoking the function if there is no stored function (i.e. operator bool returns false)
+    ///        is undefined
+    ReturnType operator()(Args... args);
```

Comment:
Exactly, this is the problem in general. It is a false guarantee but we then just terminate. Not ideal but I have no ad hoc solution for this.

Reply:
Maybe we can restrict it to noexcept callables via a user given template parameter. So that we in our use case can state that we only would like to have noexcept callables but in general the user is free to do whatever they want.

Reply:
I think it is not completely possible to check at compile time whether something can throw. I also would not like to really restrict it. Mainly because if it throws, then only because the callable you are using it with throws and the user can influence this (by providing a non throwing callable).

Reply:
Whether something can throw cannot be checked, it is not part of the signature and exceptions may originate from code the compiler does not even know when compiling the current compilation unit(other compilation unit or library, the latter being bad practice). Therefore we cannot have this. 
We can verify noexcept (which is not the same as not throwing) with the operator but I do not quite like limiting the callables. Restricting the callables to noexcept is in effect the same as declaring the operator itself noexcept. Should something be thrown, it will terminate.